### PR TITLE
py-Pillow: update to 10.0.1

### PR DESCRIPTION
**Pull request by radarhere**
###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

I realise that 10.0.1 is a security release, but the port hasn't been updated to 10.0.0 yet, and the only functional change in 10.0.1 is to update a dependency in the Python wheels - which is irrelevant to macports.

###### Tested on

macOS 13.5.2 22G91 arm64
Xcode 14.3.1 14E300c

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
```
Portfile changed since last build; discarding previous state.
--->  Computing dependencies for py-Pillow
--->  Fetching distfiles for py-Pillow
--->  Verifying checksums for py-Pillow
--->  Extracting py-Pillow
--->  Configuring py-Pillow
--->  Building py-Pillow
--->  Testing py-Pillow
```
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: python/py-Pillow/Portfile
Comment (reneeotten) [inline python/py-Pillow/Portfile:7]: version 10.0.x does not support PY37 anymore, so you will need to pin that subport to the latest supported version

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: python/py-Pillow/Portfile
Comment (radarhere) [reply python/py-Pillow/Portfile:7]: I've added a commit.

**Comment by reneeotten** [comment]
this also removed support for Qt5; so at least some due diligence needs to be done whether this update will break existing ports that depend on it.

**Comment by radarhere** [comment]
I can see that qgis3 has a 'lib' dependency on py311-pyqt5 and a 'run' dependency on py311-Pillow.

What is the solution for that? Is this blocked until pyqt6 is added to MacPorts and propagated to all ports that it depends on?

**Comment by reneeotten** [comment]
> I can see that qgis3 has a 'lib' dependency on py311-pyqt5 and a 'run' dependency on py311-Pillow.
> 
> What is the solution for that? Is this blocked until pyqt6 is added to MacPorts and propagated to all ports that it depends on? Or should I add commits to this PR to limit qgis3 and any other relevants ports to only use Pillow < 10?

There are other ports like `py-spyder` and `py-matplotlib` to name two of the ones that I maintain that use or have support for PyQt5 and have a dependency on Pillow, either through a variant or directly. There are other dependents, see for example [this page](https://ports.macports.org/port/py311-Pillow/details/). 

So yes, this is an issue that will probably take some time to get resolved. For example, I tried to add `py-pyqt6` but am running into some issues in the `destroot` phase which I haven't found the time for to look at in more detail. Pretty much the only way to right now introduce this update would be to have it as a separate port, but that is associated with other possible issues... For example, it will conflict with the "old" `py-Pillow` port and thus users will not be allowed to install these two ports or ports that depend on them simultaneously). Another question would be is how to name the ports, should the current one be renamed to `py-Pillow9` and `py-Pillow` follow the latest upstream version, or....?  

In any case, these complications will require the current maintainer @stromnov to chime in and give an opinion.

**Comment by pmetzger** [comment]
@reneeotten It appears @stromnov never chimed in.

**Comment by reneeotten** [comment]
> @reneeotten It appears @stromnov never chimed in.

this breaks too many things in its current state, so it cannot be merged. Only viable option could be to provide a legacy port pinned to v9 and have ports depend on that. That's what I said before and is all I can say about it now.

**Comment by nilason** [comment]
> So yes, this is an issue that will probably take some time to get resolved. For example, I tried to add `py-pyqt6` but am running into some issues in the `destroot` phase which I haven't found the time for to look at in more detail. Pretty much the only way to right now introduce this update would be to have it as a separate port, but that is associated with other possible issues... For example, it will conflict with the "old" `py-Pillow` port and thus users will not be allowed to install these two ports or ports that depend on them simultaneously). Another question would be is how to name the ports, should the current one be renamed to `py-Pillow9` and `py-Pillow` follow the latest upstream version, or....? Is an update to version 10 really worth the hassle at this time... for example, I would want to see first which of the dependents do allow the newer version or if most of them are still pinned to <10.

The fact that version 10 is needed for Python 3.12 support will not simplify this problem.
(I'm afraid the QGIS support of QT6 still has some distance before reaching stability to replace QT5.)

**Comment by pmetzger** [comment]
How do we either progress this or close it?

**Comment by radarhere** [comment]
If I were to update Pillow only for Python 3.12, would that be helpful?

Downside compared to this PR: It wouldn't update Pillow for earlier versions, and so isn't a complete solution.
Upside: It wouldn't break any existing functionality with Qt5, since MacPorts Pillow doesn't have a Python 3.12 subport yet.

**Comment by pmetzger** [comment]
I don't know if that's good enough. @reneeotten what are your thoughts?

**Comment by pmetzger** [comment]
@reneeotten ?

**Comment by reneeotten** [comment]
no, that will not help. We should add a py312 subport but for the version that still supports Qt5. Many packages are having/getting the py312 subport added now and then we will be in the same situation - the proposed version of py-Pillow will not work in MacPorts as we don't have support for PyQt6 yet and many packages that depend on Pillow do not support Qt6 yet either.

**Comment by pmetzger** [comment]
This doesn't seem to be making progress. Should it be closed for now?

**Comment by pmetzger** [comment]
Closing for now. If someone starts working on it again, it can be reopened.

**Comment by radarhere** [comment]
Status update
- py-pyqt6 now exists
- Matplotlib [now supports PyQt6](https://matplotlib.org/stable/gallery/user_interfaces/embedding_in_qt_sgskip.html), although the py-matplotlib port doesn't yet
- Spyder [now supports PyQt6](https://github.com/spyder-ide/spyder/blob/a672bb787c62351c1682c71d222a7221c1f11478/spyder/requirements.py#L37), although the py-spyder port doesn't yet
- QGIS [now supports PyQt6](https://changelog.qgis.org/en/entry/2610), although the qgis3 port doesn't yet

**Comment by reneeotten** [comment]
Correct, but it builds only completely/correctly on a very small number of macOSes so is not ready for prime-time yet. Contributions to improve that situation are welcome.

**Comment by radarhere** [comment]
Sorry, which port/library are you referring to in your comment?

**Comment by reneeotten** [comment]
> Sorry, which port/library are you referring to in your comment? py-pyqt6?

Yes, `py-pyqt6` is one of them but also `py-pyqt6-webengine` and in fact several of the `qt6` subports appear not to be building on anything older than macOS 15 even though upstream claims to support it. I haven't had the time to take a closer look at that, but for sure complete support for Qt6 in MacPorts is not there.

**Comment by radarhere** [comment]
> Pretty much the only way to right now introduce this update would be to have it as a separate port, but that is associated with other possible issues... For example, it will conflict with the "old" `py-Pillow` port and thus users will not be allowed to install these two ports or ports that depend on them simultaneously). Another question would be is how to name the ports, should the current one be renamed to `py-Pillow9` and `py-Pillow` follow the latest upstream version, or....?

I've created https://github.com/macports/macports-ports/pull/29472 to attempt this.

**Comment by radarhere** [comment]
I've realised there is a simple solution - just patch Qt5 support back into the latest version of Pillow. I've created https://github.com/macports/macports-ports/pull/30788

**Comment by radarhere** [comment]
#30788 has been merged.

**Pull request closed without merging**
